api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,645 @@
+app value or not; and `spill_slot_to_reg`, to denote which register’s app value a
+spill slot contains.
+
+- When a register is written by an application instruction, we invalidate all
+`spill_slot_to_reg` entries that are mapped to that register, and also set
+`gpr_is_native` for that register.
+- When a register is written by a non-drreg meta instruction, we clear
+`gpr_is_native` for that reg.
+- When a register is loaded by drreg from the slot it was spilled to, we set
+`gpr_is_native`.
+- When a register is spilled to some spill slot, we set `spill_slot_to_reg` for
+that spill slot to that reg.
+
+This strategy allows us to robustly keep track of the various corner cases that can
+arise in drreg, like spill regions from different phases overlapping (nesting or
+just overlapping), and the other known issues linked above. This was implemented by
+this
+[commit](https://github.com/DynamoRIO/dynamorio/commit/f62441bc2a0ced41263e2229deb3691433d6abb9).
+
+The drawback of this approach is that it needs to be aware of other methods of
+spilling and restoring registers outside drreg ([dropped PR](https://github.com/DynamoRIO/dynamorio/pull/4987)).
+DynamoRIO uses various such methods internally (spilling to stack, slots not
+managed by drreg), and also the client may use their own unique methods. So, some
+non-drreg meta instructions may actually restore an application value to a
+register, but this approach will not be able to recognize that. This may cause it
+to lose track of some register’s application value. We dropped this approach on
+encountering
+[DynamoRIO/dynamorio#4963](https://github.com/DynamoRIO/dynamorio/issues/4963).
+
+__Preferred: Pairing restores with spills (instead of the other way)__
+
+The key observation behind this approach is that it is easier to find the matching
+spill for a given restore, than to find the matching restore for a given spill.
+This is because there may be other restores besides the final restore, e.g.
+restores for app read, user prompted restores, etc. This makes it hard to find
+exactly where the spill region for a register/aflags ends.  Additional complexities
+include the fact that aflags re-spills may not use the same slot, which makes
+differentiating spills from multiple phases difficult.
+
+Each restore must have a matching spill. Based on this observation, we scan the
+faulting fragment from end to beginning, matching register restores to their
+spills. When we reach the faulting instruction, any restore for which we did not
+see the matching spill yet must be performed by the drreg state restoration. This
+was implemented by
+([commit](https://github.com/DynamoRIO/dynamorio/commit/be78c124eb787601182e3c73f4be8bb859c50ef8)).
+
+This algorithm does not need to be aware of non-drreg methods of spilling/restoring
+registers. Note that, like other drreg design, this method does not restore a
+spilled register's application value if it is dead at the faulting instruction.
```

Comment:
This is only true if `drreg_options_t.conservative` is set to false.  If it is true, this approach is then incorrect.  This should be documented in drreg.h and updated here.

Reply:
Pushed that code just now.

Reply:
By mistake, I used `git commit --amend` instead of `git commit`, without any only-local commit. That caused a mismatch with the origin branch. Had to manually put the new changes back.